File: programmers/단어변환.java
import java.util.*;
class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer=0;
        for (String word : words) {
			if (word.equals(target)) {
				answer = 1;
				break;
			}
		}
		if (answer != 1) {
			answer = 0;
			return 0;
		}
        answer=bfs(0,begin,target,words,0);
        
        return answer;
    }
    public int bfs(int index,String begin,String target,String[] words,int stage) {
        int answer=0;
        int pre=1;
        Queue<String> q = new LinkedList<String>();
        q.offer(begin);
        
        while (!q.isEmpty()) {
            String tmp = q.remove();
            if(tmp.equals(target)) break;
            ++stage;
            for(String s : words){
                int diff =0 ;
                for(int i=0;i<begin.length();i++){
                    if(tmp.charAt(i)!=s.charAt(i)){
                        ++diff;
                    }
                    if(diff>1){
                        break;
                    }
                }
                if(diff ==1){
                    q.offer(s);
                    ++index;
                }
            }
            if(stage == pre){
                pre = index;
                index = 0;
                stage = 0;
                ++answer;
            }
            if(answer>words.length){
                answer=0;
                break;
            }
        }
        return answer;
    }
}

File: programmers/단어변환.js
function solution(begin, target, words) {
    var answer = 0;
    let q = [begin];
    let wordArr = new Array(words.length).fill(false);
    let start = "";
    let qsize = 1;
    while(q.length>0){
        start = q.shift();
        qsize--;
        
        for(let i in words){
            if(compare(start,words[i])){
                if(wordArr[i] !=true){
                    wordArr[i] = true;
                    q.push(words[i]);
                }
                if(words[i]==target){
                    return answer+1;
                }
            }
        }
        if(qsize == 0){
            answer++;
            qsize = q.length;
        }
    }

    return 0;
}
function compare(word1, word2){
    let diff=0;
    for(let i=0;i<word1.length;i++){
        if(word1.charAt(i) !== word2.charAt(i)){
            ++diff;
        }
    }
    if(diff === 1)return true;
    return false;
}